### PR TITLE
Ports Vampire to role_datums.

### PR DESCRIPTION
**Pull request by ShiftyRail**
Finished, review points answered, ready for the final review.

### CODE REVIEW

**Comment by Probe1** [comment]
giphy

**Comment by Irradiation** [comment]
GOD DAMNIT SHIFTY, YOU HAD A CHANCE TO REMOVE VAMPIRES FROM /vg/station! 
YOU WERE MEANT TO BE THE ONE!

**Comment by ShiftyRail** [comment]
Oh no

**Comment by ShiftyRail** [comment]
Updated to do list : 
- [ ] ~Getting rid of all those `mob` level procs, all of them could be on `/datum/role/vampire` level. It would be more modular.~ It's basically a useless distinction at this point. Maybe in the future for the sake of cleanliness.
- [x] Clean up a few files here and there - I don't use the proper procs to give out objectives.
- [x] Handle de-thralling.

File: code/datums/gamemode/gamemode.dm [hunk 1]
```diff
@@ -126,8 +128,8 @@
 			i--
 			continue
 
-	return 1
-
+/datum/gamemode/proc/createBasicRole(var/type_role)
```
Comment (ShiftyRail) [inline code/datums/gamemode/gamemode.dm:131]: Requesting @MadmanMartian's take on this

File: code/datums/gamemode/role/vampire_role.dm [hunk 2]
```diff
@@ -61,11 +60,14 @@
 	for(var/type_VP in roundstart_powers)
 		var/datum/power/vampire/VP = new type_VP
 		VP.Give(src)
+	
+	faction.name = "Lord [antag.current]'s vampiric servants."
```
Comment (ShiftyRail) [inline code/datums/gamemode/role/vampire_role.dm]: Doesn't work - might change to "Vampire force #[i]"

File: code/datums/gamemode/vampire_gamemode.dm [hunk 3]
```diff
@@ -1,38 +1,77 @@
 /datum/gamemode/vampire
-	minimum_player_count = 10
+	minimum_player_count = 0 // OVERRIDE
```
Comment (ShiftyRail) [inline code/datums/gamemode/vampire_gamemode.dm]: Eh... I'll fix that

File: code/datums/gamemode/role/vampire_role.dm [hunk 4]
```diff
@@ -458,4 +445,22 @@
...
+	I.plane = VAMP_ANTAG_HUD_PLANE
+	antag.current.client.images += I
+	
+	if (new_T == src) // He just joined - let's tell him who is the master, as well as the other guys
+		var/image/I2 = image('icons/mob/mob.dmi', loc = master.antag.current, icon_state = "vampire")
+		I2.plane = VAMP_ANTAG_HUD_PLANE
+		antag.current.client.images += I2
+		for (var/datum/role/thrall/T_other in master.thralls - src)
+			var/image/I3= image('icons/mob/mob.dmi', loc = T_other.antag.current, icon_state = "vampthrall")
+			I2.plane = VAMP_ANTAG_HUD_PLANE
```
Comment (ShiftyRail) [inline code/datums/gamemode/role/vampire_role.dm]: `I3.plane`

File: __DEFINES/planes+layers.dm [hunk 5]
```diff
@@ -144,7 +144,7 @@ What is the naming convention for planes or layers?
 
 #define HUMAN_PLANE 			9			// For Humans that are standing up.
 
-//#define THIS_SPACE_FOR_RENT!	10
+#define VAMP_ANTAG_HUD_PLANE	10
```
Comment (MadmanMartian) [inline __DEFINES/planes+layers.dm:147]: This shouldn't be necessary, as the antag hud @DeityLink put in place just flicks through each icon available anyway.

File: code/datums/gamemode/factions/faction.dm [hunk 6]
```diff
@@ -135,7 +135,7 @@ var/list/factions_with_hud_icons = list()
 	return dat
 
 /datum/faction/proc/process()
-	for(var/datum/role/R in members)
+	for (var/datum/role/R in members)
```
Comment (MadmanMartian) [inline code/datums/gamemode/factions/faction.dm:138]: what the fug

File: code/datums/gamemode/factions/vampire_faction.dm [hunk 7]
```diff
@@ -2,12 +2,23 @@
 	name = "Vampire Lords"
 	desc = "Hailing from Space Transylvania."
 	ID = VAMPIRELORDS
-	required_pref = ROLE_VAMPIRE
+	required_pref = "vampire" // TO BE CHANGED TO A DEFINE
```
Comment (MadmanMartian) [inline code/datums/gamemode/factions/vampire_faction.dm]: Check why the ROLE_VAMPIRE isn't working.

File: code/datums/gamemode/role/vampire_role.dm (hunk 2 above)
Comment (MadmanMartian) [reply code/datums/gamemode/role/vampire_role.dm]: Do it in the faction.

File: code/datums/gamemode/role/vampire_role.dm [hunk 8]
```diff
@@ -166,76 +167,29 @@
 			to_chat(assailant, "<span class='notice'>You have accumulated [blood_total] [blood_total > 1 ? "units" : "unit"] of blood[blood_usable_before != blood_usable ?", and have [blood_usable] left to use." : "."]</span>")
 		check_vampire_upgrade()
 		target.vessel.remove_reagent(BLOOD,25)
+		update_vamp_hud()
```
Comment (MadmanMartian) [inline code/datums/gamemode/role/vampire_role.dm:182]: We now have a single antag hud intoduced by @DeityLink , use that.

File: code/datums/gamemode/role/vampire_role.dm [hunk 9]
```diff
@@ -166,76 +167,29 @@
...
-					to_chat(M, "[msg]")
-					antag.store_memory("<font size = 1>[msg]</font>")
+			if (!(VP.id in powers))
+				VP.Give(src)
 
+/datum/role/vampire/proc/handle_enthrall(var/datum/mind/M)
+	if (!istype(M))
+		return FALSE
+	var/datum/role/thrall/T = new(thrall = M, master = src) // Creating a new thrall
+	update_vamp_hud(T)
```
Comment (MadmanMartian) [inline code/datums/gamemode/role/vampire_role.dm]: Again, use the antag hud, not the vamp hud.

File: code/datums/gamemode/role/vampire_role.dm [hunk 10]
```diff
@@ -166,76 +167,29 @@
...
+	thralls += T
+	for (var/datum/role/thrall/T_other in thralls)
```
Comment (MadmanMartian) [inline code/datums/gamemode/role/vampire_role.dm]: The antag hud @DeityLink already made does this, use that.

File: code/datums/gamemode/role/vampire_role.dm [hunk 11]
```diff
@@ -386,7 +350,8 @@
 -- Helpers --
 */
 
-/datum/role/vampire/proc/update_vamp_hud()
+// ! - need to handle thrall removing
+/datum/role/vampire/proc/update_vamp_hud(var/datum/role/thrall/new_T = null)
```
Comment (MadmanMartian) [inline code/datums/gamemode/role/vampire_role.dm]: AAAA IT'S DEPRECIATED, STOP

File: code/datums/gamemode/role/vampire_role.dm [hunk 12]
```diff
@@ -458,4 +445,22 @@
 	antag.current << sound('sound/effects/vampire_intro.ogg')
 
 /datum/role/thrall/ForgeObjectives()
-	objectives.AddObjective(new /datum/objective/protect_master(master), src.antag)
\ No newline at end of file
+	objectives.AddObjective(new /datum/objective/protect_master(master), src.antag)
+
+/datum/role/thrall/proc/update_vamp_hud(var/datum/role/thrall/new_T = null)
```
Comment (MadmanMartian) [inline code/datums/gamemode/role/vampire_role.dm]: Again, use the antag hud, not the vamp hud.

File: code/datums/gamemode/vampire_gamemode.dm [hunk 13]
```diff
@@ -1,38 +1,77 @@
...
 	name = "Vampire"
 	factions_allowed = list(/datum/faction/vampire)
 
+	roles_allowed = 1 // Per factions
+
+	var/number_of_vamps
```
Comment (MadmanMartian) [inline code/datums/gamemode/vampire_gamemode.dm:8]: Surely there's a better way than having explicit variables per gamemode for each antag type.

File: code/datums/gamemode/vampire_gamemode.dm [hunk 14]
```diff
@@ -1,38 +1,77 @@
...
+
+/datum/gamemode/vampire/CreateFactions()
```
Comment (MadmanMartian) [inline code/datums/gamemode/vampire_gamemode.dm:10]: Same as above. There has to be a better way.

File: code/modules/spells/aoe_turf/conjure/conjure.dm [hunk 15]
```diff
@@ -23,6 +23,10 @@ How they spawn stuff is decided by behaviour vars, which are explained below
 	//should have format of list("emagged" = 1,"name" = "Wizard's Justicebot"), for example
 
 /spell/aoe_turf/conjure/cast(list/targets, mob/user)
+
+	to_chat(world, "cast()")
```
Comment (MadmanMartian) [inline code/modules/spells/aoe_turf/conjure/conjure.dm]: DEBUG TEXT, PLS REMOVE

File: code/modules/spells/aoe_turf/conjure/conjure.dm [hunk 16]
```diff
@@ -23,6 +23,10 @@ How they spawn stuff is decided by behaviour vars, which are explained below
...
+	to_chat(world, "[targets.len]")
```
Comment (MadmanMartian) [inline code/modules/spells/aoe_turf/conjure/conjure.dm]: DEBUG TEXT, PLS REMOVE

File: code/modules/spells/aoe_turf/conjure/conjure.dm [hunk 17]
```diff
@@ -63,6 +67,7 @@ How they spawn stuff is decided by behaviour vars, which are explained below
 			spawn_place.ChangeTurf(summoned_object_type)
 			summoned_object = spawn_place
 		else
+			to_chat(world, "calling summon object")
```
Comment (MadmanMartian) [inline code/modules/spells/aoe_turf/conjure/conjure.dm]: DEBUG TEXT, PLS REMOVE

File: code/modules/spells/aoe_turf/screech.dm [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+		return FALSE
+	if (!user.vampire_power(blood_cost, 0, FALSE))
+		return FALSE
+
+/spell/aoe_turf/screech/choose_targets(var/mob/user = usr)
+	
+	var/list/targets = list()
+
+	for(var/mob/living/carbon/C in hearers(user, 4))
+		to_chat(world, C)
```
Comment (MadmanMartian) [inline code/modules/spells/aoe_turf/screech.dm]: DEBUG TEXT, PLS REMOVE

File: data/mode.txt [hunk 19]
```diff
@@ -1 +1 @@
-sandbox
+Vampire
```
Comment (MadmanMartian) [inline data/mode.txt]: what the fug

**Review by MadmanMartian — CHANGES_REQUESTED** [review]

File: data/mode.txt (hunk 19 above)
Comment (ShiftyRail) [reply data/mode.txt]: It's not ready for merging yet, I just wanted a middle of the road checkpoint

File: code/datums/gamemode/role/vampire_role.dm (hunk 11 above)
Comment (DamianX) [reply code/datums/gamemode/role/vampire_role.dm]: image

File: code/datums/gamemode/vampire_gamemode.dm (hunk 14 above)
Comment (ShiftyRail) [reply code/datums/gamemode/vampire_gamemode.dm:10]: I based it on the idea of 1 vampire = 1 faction, which is quite unique to the mode and leave more room for expansion than letting all vamps be orphaned roles.

**Comment by ShiftyRail** [comment]
This commit is a work in progress, I still have things to do.

File: code/datums/gamemode/role/vampire_role.dm (hunk 8 above)
Comment (ShiftyRail) [reply code/datums/gamemode/role/vampire_role.dm:182]: This is for the blood of the vamp. The antag hud of Deity is more about who's in your team.

File: __DEFINES/planes+layers.dm (hunk 5 above)
Comment (ShiftyRail) [reply __DEFINES/planes+layers.dm:147]: I need it for the client image of the blood count.

File: code/datums/gamemode/factions/faction.dm (hunk 6 above)
Comment (ShiftyRail) [reply code/datums/gamemode/factions/faction.dm:138]: I like spaces

File: code/modules/clothing/masks/stone.dm [hunk 20]
```diff
@@ -25,6 +25,7 @@
 /obj/item/clothing/mask/stone/equipped(mob/M as mob, wear_mask)
 	if(!istype(M, /mob/living/carbon/human)) //just in case a non-human somehow manages to equip it
 		forceMove(M.loc)
+	spikes()
```
Comment (ShiftyRail) [inline code/modules/clothing/masks/stone.dm:28]: Previously in this implementation the stone mask did fuck all. Now it vampires you on equip.

**Comment by ShiftyRail** [comment]
Can you believe that ?
It's finally done
Sorry for taking so long

**Comment by Irradiation** [comment]
YOU STILL HAVE TIME TO STOP THIS, SHIFTY. BE THE MAN THAT /vg/ NEEDS.

**Review by MadmanMartian — APPROVED** [review]

**Comment by ShiftyRail** [comment]
The green tick is a good feel
There is a few things to do still though, should be done by tomorrow - mostly cleaning it up

File: code/datums/gamemode/role/vampire_role.dm (hunk 8 above)
Comment (MadmanMartian) [reply code/datums/gamemode/role/vampire_role.dm]: o ok

File: code/modules/clothing/masks/stone.dm [hunk 21]
```diff
@@ -82,3 +74,20 @@
...
+
+/proc/makeLateVampire(var/mob/living/carbon/human/H, var/blood_to_give)
+	var/datum/role/vampire/vamp =  new(H.mind, override = TRUE)
+	var/datum/faction/vampire/Fac_vamp = new(vamp)
+	if (!vamp || !Fac_vamp)
+		return FALSE
+	ticker.mode.factions += Fac_vamp
+	ticker.mode.orphaned_roles -= vamp
+	vamp.OnPostSetup()
+	log_admin("[H] has become a vampire using a stone mask.")
```
Comment (MadmanMartian) [inline code/modules/clothing/masks/stone.dm]: Doesn't strictly become a vampire using the stone mask using this proc. Move the log_admin to where it's being called in this case.

File: code/modules/clothing/masks/stone.dm [hunk 22]
```diff
@@ -82,3 +74,20 @@
 
 /obj/item/clothing/mask/stone/infinite //this mask can be used any number of times
 	infinite = 1
+
+/proc/makeLateVampire(var/mob/living/carbon/human/H, var/blood_to_give)
+	var/datum/role/vampire/vamp =  new(H.mind, override = TRUE)
```
Comment (MadmanMartian) [inline code/modules/clothing/masks/stone.dm]: There's an argument in new role for assigning a faction. Create the faction, then have it in the constructor arguments of the role. Unless the vampire factions are so bass ackwards as to go against this design.

**Review by MadmanMartian — CHANGES_REQUESTED** [review]

File: code/modules/clothing/masks/stone.dm [hunk 23]
```diff
@@ -76,18 +77,18 @@
...
 
 /proc/makeLateVampire(var/mob/living/carbon/human/H, var/blood_to_give)
-	var/datum/role/vampire/vamp =  new(H.mind, override = TRUE)
-	var/datum/faction/vampire/Fac_vamp = new(vamp)
+	var/datum/faction/vampire/Fac_vamp = new
+	var/datum/role/vampire/vamp =  new(H.mind, Fac_vamp, override = TRUE)
 	if (!vamp || !Fac_vamp)
 		return FALSE
 	ticker.mode.factions += Fac_vamp
 	ticker.mode.orphaned_roles -= vamp
```
Comment (ShiftyRail) [inline code/modules/clothing/masks/stone.dm]: This should be removed

**Review by MadmanMartian — APPROVED** [review]

**Pull request merged**